pkg/volume/util/util.go

```diff
@@ -320,62 +320,98 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
 	return pod, nil
 }
 
-// SelectZone selects a zone for a volume based on several factors:
-// node.zone, allowedTopologies, zone/zones parameters from storageclass
-// and zones with active nodes from the cluster
+// SelectZoneForVolume is a wrapper around SelectZonesForVolume
+// to select a single zone for a volume based on parameters
 func SelectZoneForVolume(zoneParameterPresent, zonesParameterPresent bool, zoneParameter string, zonesParameter, zonesWithNodes sets.String, node *v1.Node, allowedTopologies []v1.TopologySelectorTerm, pvcName string) (string, error) {
+	zones, err := SelectZonesForVolume(zoneParameterPresent, zonesParameterPresent, zoneParameter, zonesParameter, zonesWithNodes, node, allowedTopologies, pvcName, 1)
+	if err != nil {
+		return "", err
+	}
+	zone, ok := zones.PopAny()
+	if !ok {
+		return "", fmt.Errorf("could not determine a zone to provision volume in")
+	}
+	return zone, nil
+}
+
+// SelectZonesForVolume selects zones for a volume based on several factors:
+// node.zone, allowedTopologies, zone/zones parameters from storageclass,
+// zones with active nodes from the cluster. The number of zones = replicas.
+func SelectZonesForVolume(zoneParameterPresent, zonesParameterPresent bool, zoneParameter string, zonesParameter, zonesWithNodes sets.String, node *v1.Node, allowedTopologies []v1.TopologySelectorTerm, pvcName string, numReplicas uint32) (sets.String, error) {
```

Comment:
nit: Is it possible to do without some of these parameters? If we can reduce the input set, it'll help guarantee test coverage and possibly improve code maintainability

Reply:
since a lot of these parameters are mutually exclusive to each other, it might be worth having a `SelectZonesForVolume` that only takes in one of these params called `zones`, and a wrapper function that does all this logic to determine mutual exclusivity as well as picking which of the `zone|zones|topology` to pass to `SelectZonesForVolume`

Reply:
@verult The idea behind `SelectZonesForVolume` is all the in-tree providers like EBS, Azure and GCE PD should be able to parse out volume zone and topology parameters from the storage class, invoke `SelectZoneForVolume`/`SelectZonesForVolume` and get a (set of) volume(s) back to pass to the cloud provider API without having to duplicate any of the logic to enforce the rules. So `SelectZonesForVolume` is the wrapper that implements all the common logic to determine mutual exclusivity and for that it needs the full set of parameters around zone/zones/node/topology. I agree that the long list of parameters is not pretty but to help with maintenance, we have an extensive set of unit tests in `TestSelectZonesForVolume` and `TestSelectZoneForVolume`. Those should make sure none of the rules around exclusivity of the parameters are broken.

@davidz627  `SelectZonesForVolume` is the wrapper that enforces the compatibility rules among the parameters. `chooseZonesForVolumeIncludingZone`/`ChooseZonesForVolume` invoked from `SelectZonesForVolume` does the actual picking of zone(s) from a set of zones. Let me know if it makes sense and already aligns with the above suggestion.